### PR TITLE
[WIP] change set_banned and is_banned -> bool to make the ban time limited

### PR DESCRIPTION
**Pull request by sesam**
- first attempt at #439

### CODE REVIEW

File: p2p/src/peer.rs [hunk 1]
```diff
@@ -39,7 +39,7 @@ pub struct Peer {
 	pub info: PeerInfo,
 	proto: Box<Protocol>,
 	state: Arc<RwLock<State>>,
-	banned_until: Arc<RwLock<time::Tm>>,
+	banned_until: Arc<RwLock<i64>>,
```
Comment (sesam) [inline p2p/src/peer.rs]: Maybe better use a RefCell https://doc.rust-lang.org/core/cell/struct.RefCell.html

File: p2p/src/peer.rs [hunk 2]
```diff
@@ -123,16 +126,30 @@ impl Peer {
...
-		*state == State::Banned
+		*state == State::Banned && self.recheck_ban_timeout()
 	}
 
 	/// Set this peer status to banned
 	pub fn set_banned(&self) {
 		let mut state = self.state.write().unwrap();
 		*state = State::Banned;
+		let mut banned_until = self.banned_until.write().unwrap();
+		*banned_until = (time::now_utc() + time::Duration::minutes(42)).to_timespec().sec;
```
Comment (antiochp) [inline p2p/src/peer.rs]: Lol at 42 - lets do 3 hours initially? If this turns out to be too long then we can shorten it.
But I think we want to ban for an extended period for now.

Comment (sesam) [reply p2p/src/peer.rs]: Maybe wrongly, I assumed we'd just ban and ban again for nodes that aren't updated. My recent restarts only banned good nodes, so maybe leave a low number at least until that issue is fixed?

For Mainnet I'm thinking a tiered slightly random approach, and repeat offenders get bumped into lower levels of oblivion, with a slower poisson process to get kicked upwards to get another chance. When DoS attacks start, I think IP-based banning won't be effective protection, and we'll need to score connected nodes by age, trying to keep a mix of old known and recently discovered nodes in the 8 spot connection pool.

Comment (ignopeverell) [reply p2p/src/peer.rs]: What do you mean by "banned good nodes"? Nodes are banned if they send a block that will never be valid for you. If they accepted that node, there is nothing they can do to help you.

File: p2p/src/peer.rs (hunk 1 above)
Comment (sesam) [reply p2p/src/peer.rs]: (switched to RefCell and it seems I got persistence to work, yay)

File: p2p/src/peer.rs (hunk 2 above)
Comment (sesam) [reply p2p/src/peer.rs]: as per gitter chat, it seems we have examples of banning good nods incorrectly. Maybe making that painful for all the newcomers and teaching them to do rm -rf .grin/peers regularly is the right call. And mably was a bad seed node, but he already started resync and will be good again in shorter time than 3 hours.

**Comment by ignopeverell** [comment]
Sorry, looking at this until now but this approach is unlikely to work. The banned setup and time setting is fine, but then disconnected peers are regularly cleaned up from the peers list to avoid clogging up memory. So the unbanning timer will never fire. I would instead look at adding an additional check in `grin/src/seed.rs` when we try to get more peers from db.

**Comment by ignopeverell** [comment]
Any chance to have this fixed? :-)

**Comment by sesam** [comment]
#439 fixed by #631, so removing this one.

**Pull request closed without merging**